File: sdks/node/src/openai.ts
// Drop-in OpenAI client wrapper.
//
// Replace:
//   import OpenAI from "openai";
// With:
//   import { OpenAI } from "@platform/sdk/openai";

import { resolveBaseUrl } from "./routing.js";

interface OpenAIClientOptions {
  baseURL?: string;
  apiKey?: string;
  [key: string]: unknown;
}

const DIRECT_BASE_URL =
  process.env.OPENAI_BASE_URL ?? "https://api.openai.com/v1";

/**
 * Construct an OpenAI client routed through the local runtime agent
 * when reachable, otherwise the direct upstream (with a warning).
 *
 * The upstream `openai` package is a peer dependency — install it
 * separately in your application.
 */
export async function OpenAI(options: OpenAIClientOptions = {}): Promise<unknown> {
  // Peer dependency — resolved at runtime by consuming apps that
  // install `openai`. Indirect import to keep TS from requiring the
  // package at build time.
  const upstreamName = "openai";
  const upstream = await import(upstreamName);
  const baseURL = await resolveBaseUrl({
    proxyPath: "/proxy/v1",
    directDefault: DIRECT_BASE_URL,
  });
  const ClientCtor = (upstream as { default?: unknown; OpenAI?: unknown }).OpenAI
    ?? (upstream as { default?: unknown }).default;
  if (typeof ClientCtor !== "function") {
    throw new Error(
      "[platform-sdk/openai] could not locate the OpenAI constructor on the upstream `openai` package.",
    );
  }
  const Ctor = ClientCtor as new (opts: OpenAIClientOptions) => unknown;
  return new Ctor({ ...options, baseURL });
}

File: sdks/python/platform_sdk/openai.py
"""Drop-in OpenAI client wrapper.

The official ``openai`` package's ``OpenAI`` class accepts a ``base_url``
constructor arg that overrides where it sends requests. We just set
that base_url to our runtime agent's proxy path; everything else (auth,
streaming, tool calls, error types) is the original SDK.

This means you can use every feature of the upstream OpenAI SDK
(``.chat.completions.create``, ``.embeddings.create``, ``.images``,
``.batches``, etc.) without modification — only the import line changes.
"""

from __future__ import annotations

import os
from typing import Any

from platform_sdk._routing import resolve_base_url

OPENAI_DIRECT_BASE_URL = os.environ.get(
    "OPENAI_BASE_URL", "https://api.openai.com/v1"
)


class OpenAI:
    """Drop-in replacement for ``openai.OpenAI``.

    Routes through the local runtime agent when available; falls back
    to a direct upstream call (with a warning log) when not.
    """

    def __new__(cls, *args: Any, **kwargs: Any) -> Any:
        try:
            from openai import OpenAI as _UpstreamOpenAI
        except ImportError as exc:  # pragma: no cover
            raise ImportError(
                "openai package not installed. Run: pip install openai"
            ) from exc

        base_url = resolve_base_url(
            proxy_path="/proxy/v1",
            direct_default=OPENAI_DIRECT_BASE_URL,
        )
        kwargs.setdefault("base_url", base_url)
        return _UpstreamOpenAI(*args, **kwargs)


class AsyncOpenAI:
    """Async variant — same behavior, async surface."""

    def __new__(cls, *args: Any, **kwargs: Any) -> Any:
        try:
            from openai import AsyncOpenAI as _UpstreamAsyncOpenAI
        except ImportError as exc:  # pragma: no cover
            raise ImportError(
                "openai package not installed. Run: pip install openai"
            ) from exc

        base_url = resolve_base_url(
            proxy_path="/proxy/v1",
            direct_default=OPENAI_DIRECT_BASE_URL,
        )
        kwargs.setdefault("base_url", base_url)
        return _UpstreamAsyncOpenAI(*args, **kwargs)
